llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp

```diff
@@ -996,6 +989,30 @@ static Value *canonicalizeSaturatedAdd(ICmpInst *Cmp, Value *TVal, Value *FVal,
   if (!match(TVal, m_AllOnes()))
     return nullptr;
 
+  if ((Pred == ICmpInst::ICMP_UGE || Pred == ICmpInst::ICMP_UGT) &&
+      match(FVal, m_Add(m_Specific(Cmp0), m_APInt(C))) &&
+      match(Cmp1, m_SpecificIntAllowPoison(~*C))) {
```

Comment:
m_SpecificIntAllowPoison allows poison elements in splat vectors. It doesn't do anything for scalars. All of your tests appear to be scalar. So the use of `AllowPoison` isn't tested?

Reply:
I did test AllowPoison and the end result doesn't change.

Reply:
Perhaps this doesn't do anything for poison since it already folded to poison. Are poison in splat vectors just vectors of nothing but poison?

Reply:
See: uadd_sat_poison

Reply:
uadd_sat_poison is a scalar poison value. m_SpecifincIntAllowPoison won't match that.

m_SpecificIntAllowPoison allows matching a vector where some elements are poison and some are the specific constant like `<i32 31, i32 poison, i32 31, i32 31>`. For scalars it only matches the specific constant.

Reply:
Allow me to update the tests then.

Reply:
I updated the tests to reflect this!

Reply:
> uadd_sat_poison is a scalar poison value. m_SpecifincIntAllowPoison won't match that.
> 
> m_SpecificIntAllowPoison allows matching a vector where some elements are poison and some are the specific constant like `<i32 31, i32 poison, i32 31, i32 31>`. For scalars it only matches the specific constant.

I added a splat-poison vector test to show this!